### ISSUE TITLE
Can we contribute the code which has been merged in Glibc?

### ISSUE THREAD

**Issue by Yikun** [ARM-software/optimized-routines#22]
Hi, team. : )

Recently, we have contributed Kunpeng optimization patches [1] to GNU Glibc.

And we are considering whether we should integrated our patches to more repos, so as to make our optimization bring benefits to more users.

There are some question as below:
1. Can we contribute Kunpeng optimized patches (already merged in glibc master, according FSF CLA, looks like we have transform these code ownership to FSF foundation) into `optimized-routines`? Is there any legal or license problem?
2. If the  answer of # 1 is `there is no legal and license problem`, Do you think it makes any sense to contribute code to the "optimized-routines" repo?
3. Is there any synchronization strategy between the `optimized-routines` repo and upstream glibc?

[1] https://sourceware.org/git/?p=glibc.git;a=search;s=Xuelei+Zhang;st=author

cc @derekpush

**Comment by nsz-arm** [comment]
Thanks for the interest in optimized-routines. I don't have an advice right now about your existing glibc contributions. I will try to get a more complete answer here later, i just wanted to note that we will have to keep our communication public because I'm not allowed to do private conversation with  developers that can be associated with huawei (but discussions here on github or oss archived mailing lists are fine).

(Other than public discussions i think having kunpeng machines somehow available in public like the gcc compile farm would also help to deal with the barriers, assuming huawei is open to do that,
then we can verify your benchmark results)

I'm of course not against getting contributions, but so far we tried to come up with best "generic" implementations instead of separate implementations per uarch (e.g. "simd" or "sve" memcpy instead of kunpeng_memcpy), but we are open to discuss this if uarch tunings turn out to be very relevant.

**Comment by xboxfanj** [comment]
Wilco tested the generic changes on various hardware and found them to be an improvement per the glibc mailing list.

https://sourceware.org/ml/libc-alpha/2019-10/msg00665.html
https://sourceware.org/ml/libc-alpha/2019-10/msg00666.html
https://sourceware.org/ml/libc-alpha/2019-10/msg00667.html
https://sourceware.org/ml/libc-alpha/2019-10/msg00711.html

**Comment by nsz-arm** [comment]
yes generic improvements are fine. i just wanted to say that improving the generic code is preferred to adding new uarch specific implementations.

**Comment by Wilco1** [comment]
Much preferred in fact! In many cases I tried, generic code outperforms microarchitecture-specific implementations even on the targeted microarchitecture. Also as I've mentioned before, people tend to just look at microbenchmarks which repeat the same function with huge inputs many times - that is the least interesting scenario given real-world inputs are very small and always changing.

**Comment by Yikun** [comment]
> i just wanted to note that we will have to keep our communication public

@nsz-arm OK, thanks for the reminder, and I will sync these info with our team. : )

@xboxfanj thank you wirte it up, there is also a strlen_asimd improvement besides these 4 patches.

[1] https://sourceware.org/ml/libc-alpha/2019-10/msg00663.html

There is no doubt that we hope we could contribute the generic implementation in here, but we just want to sure: `Is there any legal or license problem?`

**Comment by maxim-kuvyrkov** [comment]
**I'm not a lawyer, nor do I speak for Arm**

TL;DR: It's best to contribute improvements to string/memory/math functions to optimized-routines first, and then propagate them to Glibc, Bionic, Newlib from optimized-routines.  Contributing to Glibc first makes for more work.

@Yikun , contribution agreement of optimized-routines requires copyright assignment to Arm -- specifically so that same copyright can be then transferred to FSF and patches can be merged from optimized-routines to Glibc.  However, if you already contributed a given patch to Glibc, you no longer have copyright for it, and, unfortunately, you become unable to contribute this patch to optimized-routines.

You can still contribute this patch to any open-source or proprietary project (so long as the project doesn't require copyright assignment, which you transferred to FSF) because FSF grants you license to use your patch as you please.  Specifically, Newlib and Bionic don't require copyright assignment, so they can take your patch no problem.

The caveat is that contribution to Newlib and Bionic can now be done only by yourself/Huawei, which makes more work for you.  Everyone else can use your patch only under LGPLv3 license, and that's not compatible with Newlib's and Bionic's licenses.

Hope this helps.

**Comment by nsz-arm** [comment]
thanks for the comment Maxim, i was meaning to reply that it seems we cannot take code that is already committed to glibc into optimized-routines with the current legal setup.

we plan to optimize various affected string functions, so we will have to redevelop the changes independently here. you can help that work by giving feedback or benchmark on kunpeng or other cores. i'm closing this issue now.
